### PR TITLE
nanocoap_link_format: don't drop characters in fragmented entries

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

If the directory entry returned by a CoAP server does not fit into `CONFIG_NANOCOAP_QS_MAX`, we get a call-back for the fragments.
Turns out this was not properly handled. 

### Testing procedure

I discovered this when running `aiocoap-fileserver` in `/tmp` where some very long directories live:

#### master

```
/test.sh
/.X11-unix/
/.font-unix/
/.com.google.Chrome.WsWxkT
/.ICE-unix/
/.XIM-unix/
[1]    52381 segmentation fault (core dumped)
```

#### this PR

```
> ncget coap://[fe80::607f:b1ff:fef7:689c]/
/test.sh
/.X11-unix/
/.font-unix/
/.com.google.Chrome.WsWxkT
/.ICE-unix/
/.XIM-unix/
/systemd-private-880118a9b51a45438f3c22d088d2e118-systemd-resolved.service-Jkz0l9/
/systemd-private-880118a9b51a45438f3c22d088d2e118-polkit.service-pHM9x2/
/snap-private-tmp/
/config-err-ErMZeV
/dock-replace.log
/systemd-private-880118a9b51a45438f3c22d088d2e118-fwupd.service-ygqHO4/
/systemd-private-880118a9b51a45438f3c22d088d2e118-bluetooth.service-p9wj7k/
/systemd-private-880118a9b51a45438f3c22d088d2e118-systemd-timesyncd.service-ULRs39/
/systemd-private-880118a9b51a45438f3c22d088d2e118-upower.service-D4dzsc/
/.com.google.Chrome.kk8kyD/
/systemd-private-880118a9b51a45438f3c22d088d2e118-colord.service-ChuMux/
/main.c
/systemd-private-880118a9b51a45438f3c22d088d2e118-ModemManager.service-aWDcAc/
/.X0-lock
/systemd-private-880118a9b51a45438f3c22d088d2e118-systemd-logind.service-HQ3Vxw/
```

### CODE REVIEW

File: sys/net/application_layer/nanocoap/link_format.c [hunk 1]
```diff
@@ -58,16 +58,20 @@ static int _dirlist_cb(void *arg, size_t offset, uint8_t *buf, size_t len, int m
             }
         }
 
-        if (*c == ',' || ctx->cur == ctx->end) {
-            int res;
+        bool found_end = false;
+        if (*c == ',') {
+            found_end = true;
+        } else {
```
Comment (maribu) [inline sys/net/application_layer/nanocoap/link_format.c:64]:
```suggestion
        }
        else {
```

File: sys/shell/cmds/nanocoap_vfs.c [hunk 2]
```diff
@@ -52,16 +52,30 @@ static bool _is_dir(const char *url)
...
+        return 0;
+    }
+    entry = start + 1;
+
+find_end:
+    end = strchr(entry, '>');
+    if (end == NULL) {
+        *too_long = true;
+        printf("%s", entry);
+    } else {
```
Comment (maribu) [inline sys/shell/cmds/nanocoap_vfs.c:73]:
```suggestion
    }
    else {
```

File: sys/shell/cmds/nanocoap_vfs.c [hunk 3]
```diff
@@ -52,16 +52,30 @@ static bool _is_dir(const char *url)
...
+    if (start == NULL) {
+        return 0;
+    }
+    entry = start + 1;
+
+find_end:
+    end = strchr(entry, '>');
+    if (end == NULL) {
+        *too_long = true;
+        printf("%s", entry);
```
Comment (maribu) [inline sys/shell/cmds/nanocoap_vfs.c:72]: Is `entry` `\0` terminated here?

**Review by maribu — COMMENTED** [review]
thx for addressing this :)

File: sys/shell/cmds/nanocoap_vfs.c (hunk 3 above)
Comment (benpicco) [reply sys/shell/cmds/nanocoap_vfs.c:72]: Yes, `nanocoap_link_format_get_url()` will take care of this.

**Review by maribu — APPROVED** [review]

**Pull request merged**